backend/src/database/services/base-service.ts

```diff
@@ -0,0 +1,172 @@
+        manager: EntityManager
+    ): Promise<TEntity> | TEntity;
+
+    protected abstract readonly entityTarget: EntityTarget<TEntity>;
+
+    public async create(
+        creator: Creatable<TEntity>,
+        entityManager?: EntityManager
+    ): Promise<TEntity> {
+        const create = async (manager: EntityManager) =>
+            manager.save(
+                await this.createSavableObject(
+                    undefined,
+                    // TODO: TypeScript does strange things (@Dassderdie)
+                    creator as any,
+                    manager
+                )
+            );
+        return entityManager
+            ? create(entityManager)
+            : this.dataSource.transaction(create);
+    }
+
+    public async findAll(
+        options?: FindManyOptions<TEntity>,
+        entityManager?: EntityManager
+    ): Promise<TEntity[]> {
+        const find = async (manager: EntityManager) =>
+            manager.find(this.entityTarget, options);
+        return entityManager
+            ? find(entityManager)
+            : this.dataSource.transaction(find);
+    }
+
+    /**
+     * Finds the first row matching the {@link options}.
+     *
+     * @returns The found row, or `null` if no matching row was found and {@link mustExist} is `false`.
+     * @throws {@link DatabaseError} when {@link mustExist} is `true` and no row has been found.
+     */
+    public async findOne<TExists extends false | true>(
+        options: FindOneOptions<TEntity>,
+        mustExist: TExists,
+        entityManager?: EntityManager
+    ): Promise<TExists extends true ? TEntity : TEntity | null> {
+        const find = async (manager: EntityManager) => {
+            const tuple = await manager.findOne(this.entityTarget, options);
+            if (mustExist && tuple === null) {
+                throw new DatabaseError(
+                    `\`${
+                        this.entityTarget.constructor.name
+                    }\` with options \`${JSON.stringify(
+                        options
+                    )}\` could not be found.`
+                );
+            }
+            return tuple;
+        };
+        // TypeScript can't get the correct typings for this.
+        return (
+            entityManager
+                ? find(entityManager)
+                : this.dataSource.transaction(find)
+        ) as Promise<TExists extends true ? TEntity : TEntity | null>;
+    }
+
+    /**
+     * Finds the first row matching the {@link options} and returns it.
+     * If no row matches uses {@link creatable} to create a new row and return this row.
+     */
+    public async findOneOrCreate(
+        options: FindOneOptions<TEntity>,
+        creatable: Creatable<TEntity>,
+        entityManager?: EntityManager
+    ): Promise<TEntity> {
+        const find = async (manager: EntityManager) =>
+            (await this.findOne(options, false, manager)) ??
+            (await this.create(creatable, manager));
```

Comment:
Are you sure this'll work? It needs to call `create` if and only if `findOne` returns `null`.

Reply:
Oh, I see.

Then this would be equal:

```suggestion
        const find = async (manager: EntityManager) =>
            (await this.findOne(options, false, manager)) ??
            this.create(creatable, manager);
```

But the current code looks better because it only uses one concept.